src/frontend/src/handler/drop_sink.rs

```diff
@@ -64,67 +64,60 @@ pub async fn handle_drop_sink(
-                    let table = reader.get_any_table_by_id(target_table_id)?;
-                    table.clone()
-                };
-
-                let (mut graph, mut table, source, target_job_type) =
-                    reparse_table_for_sink(&session, &table_catalog).await?;
-
-                assert!(!table_catalog.incoming_sinks.is_empty());
-
-                table
-                    .incoming_sinks
-                    .clone_from(&table_catalog.incoming_sinks);
-
-                let mut incoming_sink_ids: HashSet<_> =
-                    table_catalog.incoming_sinks.iter().copied().collect();
-
-                assert!(incoming_sink_ids.remove(&sink_id.sink_id));
-
-                let columns_without_rw_timestamp = table_catalog.columns_without_rw_timestamp();
-                for sink in fetch_incoming_sinks(&session, &incoming_sink_ids)? {
-                    hijack_merger_for_target_table(
-                        &mut graph,
-                        &columns_without_rw_timestamp,
-                        &sink,
-                        Some(&sink.unique_identity()),
-                    )?;
-                }
+    if sink.stream_job_status == StreamJobStatus::Creating && sink.target_table.is_none() {
+        let canceled_jobs = session
+            .env()
+            .meta_client()
+            .cancel_creating_jobs(PbJobs::Ids(CreatingJobIds {
+                job_ids: vec![sink_id.sink_id],
+            }))
+            .await?;
+        tracing::info!(?canceled_jobs, "cancelled creating jobs");
+    } else {
+        // No matter whether the sink is being created or not, we need to replace the target table graph.
+        // TODO(august): using cancel interface to cancel the creating job after refactoring sink into table.
+        let mut affected_table_change = None;
+        if let Some(target_table_id) = &sink.target_table {
+            let table_catalog = {
+                let reader = session.env().catalog_reader().read_guard();
+                let table = reader.get_any_table_by_id(target_table_id)?;
+                table.clone()
+            };
 
-                affected_table_change = Some(ReplaceJobPlan {
-                    replace_job: Some(replace_job_plan::ReplaceJob::ReplaceTable(
-                        replace_job_plan::ReplaceTable {
-                            table: Some(table.to_prost()),
-                            source: source.map(|x| x.to_prost()),
-                            job_type: target_job_type as _,
-                        },
-                    )),
-                    fragment_graph: Some(graph),
-                });
+            let (mut graph, mut table, source, target_job_type) =
+                reparse_table_for_sink(&session, &table_catalog).await?;
+
+            let mut incoming_sink_ids: HashSet<_> =
+                table_catalog.incoming_sinks.iter().copied().collect();
+            incoming_sink_ids.remove(&sink_id.sink_id);
```

Comment:
Should we keep the assertion?

Reply:
Nope, the update of the table catalog and the synchronization of the sink catalog are not concurrent. The table catalog update occurs after the barrier collection. Therefore, when dropping a creating sink, the table's `incoming_sinks` may not yet be updated. Will add it back when the refactoring is done.

Reply:
Seems dangerous to have inconsistent metadata on frontend side. Do we plan to address the inconsistency in near future via notification with more atomicity?